### DISCUSSION TITLE
modify body in middleware

### DISCUSSION THREAD

**Discussion by piquu** [tokio-rs/axum#3002] [Q&A] [answered]
### Summary

im trying to make a middleware that wraps the body in a template. i tried some solutions from stackoverflow and github issues but i cant get it to work properly. 
```rust
pub async fn inject_base(request: Request, next: Next) -> Response {
    let response = next.run(request).await;
    let inject = !response.headers().contains_key("HX-Request");

    let (parts, body) = response.into_parts();
    let body = body::to_bytes(body, usize::MAX).await.unwrap();
    let body = String::from_utf8(body.to_vec()).unwrap();
    let body = Body::from(if inject {
        BaseTemplate {
            title: "",
            style: "",
            body,
        }.to_string()
    } else {
        body
    });

    Response::from_parts(parts, body)
}
```
the error im getting
```
payload claims content-length of 333, custom content-length header claims 56
```

### axum version

0.7.7

**Comment by yanns** [comment, marked as answer]
This was fixed in https://github.com/tokio-rs/axum/pull/2897 that is not released yet

**Comment by selfhoster1312** [comment]
Not sure how or why, but i just stumbled upon this issue:

```
thread 'main' (804778) panicked at /home/user/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-1.8.1/src/proto/h1/role.rs:704:41:
payload claims content-length of 3707, custom content-length header claims 3719
```

Strangely, it only happens on a single route, and i don't think i've modified much since it last worked. I'm sure i haven't touched the middlewares. I'll try to isolate the problem.

EDIT: for posterity's sake, it looks like the problem was using `axum::routing::get` twice around the same handler. How did that happen? I had an `add_routes` method somewhere, which did some `router.route('whatever', get(Self::whatever_handler))` where `fn whatever_handler() -> MethodRouter<State, Infallible>`. Apparently `get` does not modify the signature so the associated function returning `get(actual_handler)`  was only caught at runtime.

**Reply by mladedav** [reply]
`router.route("/", get(get(|| async {})))` should just work, I think the issue had to be something else, maybe you were still using the old version of axum where this issue was not fixed? And maybe you have some compression layers or something like that?

If you are able to reproduce it and you're on 0.8.x (preferably the latest version), please feel free to open a new bug report.
